src/vocabulary/mod.rs

```diff
@@ -0,0 +1,308 @@
+        Vocabulary::new(None).extend(tokens_and_ids)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn insert() {
+        let vocabulary = Vocabulary::new(None)
+            .insert("blah", 0)
+            .insert("1a", 1)
+            .insert("2", 2)
+            .insert("0", 3);
+
+        assert_eq!(vocabulary.len(), 4);
+        assert_eq!(vocabulary["blah"], &[0]);
+        assert_eq!(vocabulary["1a"], &[1]);
+        assert_eq!(vocabulary["2"], &[2]);
+        assert_eq!(vocabulary["0"], &[3]);
+    }
+
+    #[test]
+    fn extend() {
+        let vocabulary = Vocabulary::new(None).extend([
+            ("blah", vec![0]),
+            ("1a", vec![1]),
+            ("2", vec![2]),
+            ("0", vec![3]),
+        ]);
+
+        assert_eq!(vocabulary.len(), 4);
+        assert_eq!(vocabulary["blah"], &[0]);
+        assert_eq!(vocabulary["1a"], &[1]);
+        assert_eq!(vocabulary["2"], &[2]);
+        assert_eq!(vocabulary["0"], &[3]);
+    }
+
+    #[test]
+    fn pretrained_from_gpt2() {
+        let model = "openai-community/gpt2";
+        let tokenizer = Tokenizer::from_pretrained(model, None).expect("Tokenizer failed");
+        let vocabulary = Vocabulary::from_pretrained(model, None).expect("Vocabulary failed");
+
+        let v_eos = vocabulary.eos_token_id.expect("No eos token in vocabulary");
+        assert_eq!(v_eos, 50256);
+        assert_eq!(
+            tokenizer.id_to_token(v_eos).expect("Token not found"),
+            "<|endoftext|>"
+        );
+
+        let token = "Ġal";
+        assert!(vocabulary.token_to_ids(token).is_none());
+        assert!(tokenizer.token_to_id(token).is_some());
+
+        for (v_token, t_token_expected) in [("abc", "abc"), (" O", "ĠO")] {
+            let v_ids = vocabulary.token_to_ids(v_token);
+            assert!(v_ids.is_some());
+            for v_id in v_ids.unwrap() {
+                let t_token = tokenizer
+                    .id_to_token(*v_id)
+                    .expect("Token id not found in tokenizer");
+                assert_eq!(&t_token, t_token_expected);
+            }
+        }
+    }
+
+    #[test]
+    fn pretrained_from_llama() {
+        let model = "hf-internal-testing/llama-tokenizer";
+        let tokenizer = Tokenizer::from_pretrained(model, None).expect("Tokenizer failed");
+        let vocabulary = Vocabulary::from_pretrained(model, None).expect("Vocabulary failed");
+
+        let v_eos = vocabulary.eos_token_id.expect("No eos token in vocabulary");
+        assert_eq!(v_eos, 2);
+        assert_eq!(
+            tokenizer.id_to_token(v_eos).expect("Token not found"),
+            "</s>"
+        );
+
+        for (v_token, t_token_expected) in [
+            ("abc", "abc"),
+            (" al", "▁al"),
+            (" O", "▁O"),
+            ("   ", "▁▁▁"),
+            // TODO: won't pass since first we need to change token's type to bytes
+            // ("<0xFF>", "ÿ"),
+            // ("<0x20>", "▁"),
```

Comment:
Is this for another PR? Seems essential.

Reply:
It is essential, but without the interface change of the token type from String to bytes it won't (and not supposed to) pass. I will be addressing existing interface change of the vocabulary in the following PRs. Also, thus the question n.1 from the above.